src/ipc/ipc3/helper.c

```diff
@@ -134,16 +139,23 @@ static const struct comp_driver *get_drv(struct sof_ipc_comp *comp)
 		}
 	}
 
-	if (!drv)
+	k_spin_unlock(&drivers->lock, key);
+
+	if (!drv) {
 		tr_err(&comp_tr,
 		       "get_drv(): the provided UUID (%8x%8x%8x%8x) doesn't match to any driver!",
 		       *(uint32_t *)(&comp_ext->uuid[0]),
 		       *(uint32_t *)(&comp_ext->uuid[4]),
 		       *(uint32_t *)(&comp_ext->uuid[8]),
 		       *(uint32_t *)(&comp_ext->uuid[12]));
+		goto out;
+	}
 
-	k_spin_unlock(&drivers->lock, key);
-
+	if (drv->type != comp->type && drv->type != SOF_COMP_MODULE_ADAPTER) {
```

Comment:
what exactly is this condition: `drv->type == SOF_COMP_MODULE_ADAPTER`? So it's possible to have a component in topology with UUID, matching `SOF_COMP_MODULE_ADAPTER` but with a different type? Or how is this possible?

Reply:
I see what you are doing here, but it doesn't validate that the comp type is the right type. E.g. this still leaves the possibility for a SRC UUID with a VOL ENUM.

@lyakh how terrible would it be if we just moved that busted switch case out to each component and just `container_of` our way back to the correct pointer for vol and src?

Reply:
For instance MUX components use module adapter. I could not figure out how dig out what is hidden behind the module adapter framework. The nature of IPC3 is still quite strange to me.

Reply:
bleh, my `container_of` approach won't work since the data is parsed and copied, not aliased. 

Honestly I am more and more inclined to just suggest we go back to #7830 (with an additional guard for vol) and move that check into a IPC3 only block. There is not easy way fix the protocol without making a major revision (deprecating the enum field) and fixing legacy component init payloads. Also if we ifdef it around IPC3 it will be very clear it can be deleted when IPC3 is removed from the repo.

Reply:
we can try to modify module_adapter.c to resolve this issue.

Reply:
@cujomalainey @jsarha it seems to me that we just have to go with the "first principles" here: if we don't trust what we get over IPC, then we need to check everything. E.g. `.type` is checked in `comp_specific_builder()` but unrecognised type is just silently ignored and that doesn't seem right to me. Because if we do happen to have `comp->ext_data_length != 0` then `get_drv()` might use UUID to get a matching driver and we'll continue with an invalid type.
Then we also get size from those topology IPCs. And in general it's valid for module-adapter components to have a size of 0. In that case no initialisation data is allocated. So individual drivers must check for that, yes.